### DISCUSSION TITLE
Alternative way of detecting SemanticDB

### DISCUSSION THREAD

**Discussion by ckipp01** [scalameta/metals#4290] [General]
## Outline

So recently @lefou [merged in](https://github.com/com-lihaoyi/mill/pull/1977) much better SemanticDB support for Mill when using Mill as a build server. This means Metals users can now use Mill as a BSP server (in the upcoming release) without any manual changes needed to their build. SemanticDB related features work:

![2022-08-23 08 56 13](https://user-images.githubusercontent.com/13974112/186091421-030c9101-d930-4fd3-9e22-954b80287f3c.gif)

The one issue though that you'll hit on is that we fully rely on scanning the `scalacOptions` to see if SemanticDB is enabled:

_Here for Scala_

https://github.com/scalameta/metals/blob/18192e36d662f1947d47d135c3361783aed21e13/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L873-L884

_Here for Java_

https://github.com/scalameta/metals/blob/18192e36d662f1947d47d135c3361783aed21e13/metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala#L809-L813

Mill takes a bit different approach than the other build servers to in that there's a separate task that creates the SemanticDB only when a tool like Metals is detected. Since it's a separate task you won't actually see the thing we look for in the `scalacOptions`, but you sill find SemanticDB in the `classDirectory`.

```
"classDirectory": "file:///Users/ckipp/Documents/scala-workspace/minimal/out/minimal/compiledClassesAndSemanticDbFiles.dest"
```

Given this situation we don't want to tell the user that there is no SemanticDB when there is. So how should we handle this situation?

### Possible Solutions
 
1. We detect that `mill-bsp` is being used and if it's a new enough version we just _assume_ that it is?
2. We detect that `mill-bsp` is being used and then doublecheck that the `classDirectory` name contains `SemanticDb`?
3. We use `data` during the initialize and introduce some `semanticDbOverride: true` or something like that which notifies the client that SemanticDB is being produced even though we don't see it reflected in the `scalacOptions`?
4. Something else?

Before digging into this I'm curious if anyone has any thoughts on the above or potential solutions for this.

**Comment by tgodzik** [comment]
I think 1. should be enough together with maybe semanticdb being reported as missing similar to compilation status? However, if we plan on more tools working with that approach, maybe sending some data to Metals with the info where semanticdb might be located would probably better.

**Reply by lefou** [reply]
Mill detects the fact that a BSP client wants SemanticDB data at initialization time. We parse the extra data and read the requested SemanticDB version. Unfortunately, the `InitializeBuildResult` from the BSP protocol has no back channel, where we can push some details to the client.

**Comment by ckipp01** [comment]
I created https://github.com/scalameta/metals/pull/4295 to partially address this. This will get rid of the warnings for `mill-bsp` users about semanticdb and sourceroot not being set for Scala targets but they will still get a warning about Java support.
